services-core/src/gas_price/gnosis_safe.rs

```diff
@@ -0,0 +1,157 @@
+//! Gnosis Safe gas station `GasPriceEstimating` implementation.
+//! Api documentation at https://safe-relay.gnosis.io/ .
+
+use super::{linear_interpolation, GasPriceEstimating};
+use crate::http::{HttpClient, HttpFactory, HttpLabel};
+use anyhow::Result;
+use ethcontract::U256;
+use isahc::http::uri::Uri;
+use pricegraph::num;
+use serde::Deserialize;
+use serde_with::rust::display_fromstr;
+use std::{convert::TryInto, time::Duration};
+
+/// The default uris at which the gas station api is available under.
+const DEFAULT_MAINNET_URI: &str = "https://safe-relay.gnosis.io/api/v1/gas-station/";
+const DEFAULT_RINKEBY_URI: &str = "https://safe-relay.rinkeby.gnosis.io/api/v1/gas-station/";
+
+pub fn api_url_from_network_id(network_id: &str) -> Option<&'static str> {
+    match network_id {
+        "1" => Some(DEFAULT_MAINNET_URI),
+        "4" => Some(DEFAULT_RINKEBY_URI),
+        _ => None,
+    }
+}
+
+/// Gas prices in wei retrieved from the gas station. This is a result from the
+/// API call.
+#[derive(Deserialize, Debug, Default)]
+#[serde(rename_all = "camelCase")]
+pub struct GasPrices {
+    pub last_update: String,
+    #[serde(with = "display_fromstr")]
+    pub lowest: f64,
+    #[serde(with = "display_fromstr")]
+    pub safe_low: f64,
+    #[serde(with = "display_fromstr")]
+    pub standard: f64,
+    #[serde(with = "display_fromstr")]
+    pub fast: f64,
+    #[serde(with = "display_fromstr")]
+    pub fastest: f64,
+}
+
+const FAST_PERCENTILE: f64 = 0.75;
+const STANDARD_PERCENTILE: f64 = 0.5;
+const SAFE_LOW_PERCENTILE: f64 = 0.3;
+
+const SECONDS_PER_BLOCK: f64 = 15.0;
+// Treat percentiles as probabilities for geometric distribution.
+const FAST_TIME: f64 = SECONDS_PER_BLOCK * 1.0 / FAST_PERCENTILE;
+const STANDARD_TIME: f64 = SECONDS_PER_BLOCK * 1.0 / STANDARD_PERCENTILE;
+const SAFE_LOW_TIME: f64 = SECONDS_PER_BLOCK * 1.0 / SAFE_LOW_PERCENTILE;
```

Comment:
Can you comment with a bit more context why the expected time is seconds per block divided by percentile and why a geometric distribution is reasonable? Also is the `* 1.0` needed.

@Uxio0 might also have some thoughts here on how the safe gas estimate levels translate into expected execution times.

Reply:
I added a comment to the code to explain more. Let me know in this thread if you have more questions.

// For this module we need to estimate confirmation times. So we need some way of converting the
// percentiles into time. The standard percentile is 50% which means transactions at this gas price
// were included in 50% of the blocks. Thus for every block a transaction at this price has a 0.5
// chance to be included. We treat this as a geometric distribution in which the expected time for
// the event to happen (transaction is included in block) is 1/p. Blocks happen every 15 seconds so
// this is how we get a time estimate.
// In reality this estimation can be problematic when gas prices are changing quickly. When prices
// rising, the prices calculated based on the last 200 lagg behind the real price. For this reason
// we insert two extra points for the linear interpolation below in estimate_limits.
// We do not make use of the lowest and fastest gas price because they are too strong outliers. The
// lowest gas price is skewed by miners including their own transactions at 1 gwei. The highest gas
// price can be 1000 times the fast gas price which is not reasonable.
// I suspect this happens when people try to ensure that two of their own transactions are included
// together in the same block by picking a low gas price for the first nonce and a very high one for
// the second. Or maybe just someone making in the wei/gwei unit conversion.

Reply:
Sorry @fleupold , I've just read that. When the gas station was built we tried to estimate the execution times, but at that time even eth gas station was not doing a good estimation, so @denisgranha and me decided to leave it out as it was not important for the Safe use case

Reply:
hey @fleupold the percentils were taken based on trial and error experience at that time, and trying to make it simple. We took some codebase as template I don't remember which 

If you want to do it properly, you would need to take into account mempool, but mempool compared to the main chain it's tricky because transactions might not go through your mempool, it depends on your peer and how are you connected so you would need a cluster of nodes with different peers/locations in order to ensure proper sample size